doc/sphinx/proof-engine/ltac.rst

```diff
@@ -29,159 +30,95 @@ Syntax
+The constructs in :token:`ltac_expr` are :term:`left associative`.
 
-.. insertprodn term_ltac term_ltac
+.. insertprodn ltac_expr tactic_atom
 
 .. prodn::
-   term_ltac ::= ltac : ( @ltac_expr )
+   ltac_expr ::= {| @ltac_expr4 | @binder_tactic }
+   ltac_expr4 ::= @ltac_expr3 ; {| @ltac_expr3 | @binder_tactic }
+   | @ltac_expr3 ; [ @for_each_goal ]
+   | @ltac_expr3
+   ltac_expr3 ::= @l3_tactic
+   | @ltac_expr2
+   ltac_expr2 ::= @ltac_expr + {| @ltac_expr2 | @binder_tactic }
+   | @ltac_expr || {| @ltac_expr2 | @binder_tactic }
+   | @l2_tactic
+   | @ltac_expr1
+   ltac_expr1 ::= @value_tactic
+   | @qualid {* @tactic_arg }
+   | @l1_tactic
+   | @ltac_expr0
+   tactic_arg ::= @value_tactic
+   | @term
+   | ()
+   ltac_expr0 ::= ( @ltac_expr )
+   | [> @for_each_goal ]
+   | @tactic_atom
+   tactic_atom ::= @int
+   | @qualid
+   | ()
+
+.. todo PR For the moment, I've left the language constructs like +, || and ; unchanged in the grammar.
+   Not sure what to do with them.  If we just make these indirections I think the grammar no longer
+   gives you an overall idea of the concrete grammar without following the hyperlinks for many terms--not so easy
+   (e.g. I have a construct and I want to figure out which productions generate it so I can read about them).
+   We should think about eventually having a cheat sheet for the constructs, perhaps as part of the
+   chapter introduction (use case: I know there's a construct but I can't remember its syntax).  They
+   do show up in the index but they're not so easy to find.  I had thought a little about putting
+   an ltac expression cheat sheet at the top of the tactics index.  Unconventional, but people would
+   see it and remember how to find it.
+
+   OTOH, as you rightly note, they are not really tactics.  Looking for better ideas that we are OK with.
+
+.. note::
+
+   Syntactically, almost all tactics are :token:`l1_tactic`\s.
+   The documentation also describes the following as tactics because they can
+   generally be used anywhere a tactic can even though they fit into other levels
+   of :token:`ltac_expr`:
+
+   - :production:`binder_tactic`\s are: :tacn:`fun`, :tacn:`let` and :tacn:`info`
+   - :production:`l3_tactic`\s include |Ltac| tactics: :tacn:`try`,
+     :tacn:`do`, :tacn:`repeat`, :tacn:`timeout`, :tacn:`time`, :tacn:`progress`, :tacn:`once`,
+     :tacn:`exactly_once`, :tacn:`only`, :tacn:`abstract` and :tacn:`infoH`
+   - :production:`l2_tactic`\s are: :tacn:`tryif`
+   - :production:`l1_tactic`\s are: :tacn:`first`, :tacn:`solve`, :tacn:`idtac`, :tacn:`fail` and
+     :tacn:`gfail` as well as :tacn:`match`, :tacn:`match goal` and their :n:`lazymatch` and
+     :n:`multimatch` variants.
+   - :production:`value_tactic`\s, which return values rather than change the proof state.
+     They are: :tacn:`eval`, :tacn:`context`, :tacn:`numgoals`, :tacn:`fresh`, :tacn:`type of`
+     and :tacn:`type_term`.
+
+.. note::
+
+   - The grammar reserves the token ``||``.
```

Comment:
The way you've integrated this change will rebasing looks quite odd. I think the note should be below the one on precedence, which you cut in the middle.